kythe/rust/extractor/src/lib.rs

```diff
@@ -0,0 +1,104 @@
+// Copyright 2020 The Kythe Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#![feature(rustc_private)]
+
+extern crate rustc_driver;
+extern crate rustc_interface;
+extern crate rustc_save_analysis;
+extern crate rustc_session;
+
+use rustc_driver::run_compiler;
+use rustc_driver::{Callbacks, Compilation};
+use rustc_interface::{interface, Queries};
+use rustc_save_analysis::DumpHandler;
+use std::path::PathBuf;
+
+/// Generate a save_analysis in `output_dir`
+///
+/// `rustc_arguments` is a `Vec<String>` containing Rust compiler arguments. The
+/// first element must be an empty string.
+/// The save_analysis JSON output file will be located at
+/// {output_dir}/save-analysis/{crate_name}.json
+pub fn generate_analysis(rustc_arguments: Vec<String>, output_dir: PathBuf) -> Result<(), String> {
```

Comment:
I missed this before, but the only analysis the extractor should be doing is determining what files are necessary for the build.  The Java and C++ extractors run the corresponding compiler for related reasons.  In Java's case the naive estimate of required files tends to vastly overestimate what's actually required, so it runs the compiler to determine the set of files actually inspected during compilation (and handle plugins/annotation processors).  C++ is similar and we need to run the preprocessor to actually determine the necessary files.

This might be true for Rust, but we should *not* be doing any more analysis than that.